### PR TITLE
Adding build support for the ModalAI VOXL 2 platform

### PR DESCRIPTION
**Pull request by katzfey**
Please use [PX4 Discuss](http://discuss.px4.io/) or [Slack](http://slack.px4.io/) to align on pull requests if necessary. You can then open draft pull requests to get early feedback.

## Describe problem solved by this pull request

This is the first pull request to add support for the ModalAI VOXL 2 single board computer.

## Describe your solution

This is the bare minimum of code to test the build environment for VOXL 2

## Test data / coverage

A simple "Hello World" message output shows that PX4 is running

### CODE REVIEW

File: cmake/px4_add_library.cmake [hunk 1]
```diff
@@ -44,7 +44,11 @@ function(px4_add_library target)
 	target_compile_definitions(${target} PRIVATE MODULE_NAME="${target}")
 
 	# all PX4 libraries have access to parameters and uORB
-	add_dependencies(${target} uorb_headers parameters)
+	# TODO: Exclusion of qurt is temporary until these modules
+	#       build cleanly.
+	if(NOT ${PLATFORM} MATCHES "qurt")
+		add_dependencies(${target} uorb_headers parameters)
```
Comment (dagar) [inline cmake/px4_add_library.cmake:50]: We might be able to remove this entirely.

Comment (katzfey) [reply cmake/px4_add_library.cmake:50]: Should I remove it now or are you going to do some more research?

File: cmake/kconfig.cmake [hunk 2]
```diff
@@ -228,6 +228,14 @@ if(EXISTS ${BOARD_DEFCONFIG})
 
         # platform-specific include path
         include_directories(${PX4_SOURCE_DIR}/platforms/${PX4_PLATFORM}/src/px4/common/include)
+
+        if("${PX4_BOARD}" MATCHES "modalai_voxl2")
```
Comment (dagar) [inline cmake/kconfig.cmake:232]: To be honest this inclusion is pretty bad.

At a minimum could it be moved to `boards/modalai/voxl2/CMakeLists.txt` which is already included automatically?

I would also think most (but probably not all) of what's in voxl2_posix.cmake and voxl2_qurt.cmake really belongs under platforms/ with appropriate kconfig options for things like ORB_COMMUNICATOR, CONFIG_PARAM_SERVER, etc. We can figure this out incrementally though.

File: boards/modalai/voxl2/cmake/voxl2_posix.cmake [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+)
+
+# set(DISABLE_PARAMS_MODULE_SCOPING TRUE)
+
+add_definitions(-DORB_COMMUNICATOR)
+add_definitions(-DRELEASE_BUILD)
+
+set(CONFIG_PARAM_SERVER "1")
+
+add_compile_options($<$<COMPILE_LANGUAGE:C>:-std=gnu99>)
```
Comment (dagar) [inline boards/modalai/voxl2/cmake/voxl2_posix.cmake:53]: Why is this necessary? Isn't it already being set project wide?

File: boards/modalai/voxl2/cmake/voxl2_posix.cmake [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-std=gnu++14>)
+
+add_compile_options( -Wno-array-bounds )
```
Comment (dagar) [inline boards/modalai/voxl2/cmake/voxl2_posix.cmake:56]: Probably better to push this option into the specific module, library, etc.

File: boards/modalai/voxl2/cmake/voxl2_posix.cmake (hunk 3 above)
Comment (katzfey) [reply boards/modalai/voxl2/cmake/voxl2_posix.cmake:53]: You mean the gnu99 part?

File: boards/modalai/voxl2/cmake/voxl2_posix.cmake [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+add_definitions( -D__PX4_LINUX )
+
+link_directories(/home ${PX4_SOURCE_DIR}/boards/modalai/voxl2/lib)
+
+include(CMakeParseArguments)
+
+# Process Apps proc app source and libs
+function (LINUX_APP)
```
Comment (dagar) [inline boards/modalai/voxl2/cmake/voxl2_posix.cmake:65]: I'm not seeing anything special in this LINUX_APP(), did you try using the regular add_executable(px4) that the other linux builds use?

File: boards/modalai/voxl2/cmake/voxl2_qurt.cmake [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+	"${PX4_SOURCE_DIR}/platforms/qurt/cmake"
+)
+
+include(Toolchain-qurt)
+include(qurt_reqs)
+
+include_directories(${HEXAGON_SDK_INCLUDES})
+
+add_definitions(-DORB_COMMUNICATOR)
+add_definitions(-DRELEASE_BUILD)
```
Comment (dagar) [inline boards/modalai/voxl2/cmake/voxl2_qurt.cmake:40]: Why are you setting RELEASE_BUILD manually?

Comment (katzfey) [reply boards/modalai/voxl2/cmake/voxl2_qurt.cmake:40]: This is only to control log messages and shouldn't be there. Will remove it for both default and qurt.

File: boards/modalai/voxl2/cmake/voxl2_posix.cmake (hunk 4 above)
Comment (katzfey) [reply boards/modalai/voxl2/cmake/voxl2_posix.cmake:56]: Doesn't look like this is currently required so I'll remove. If it is needed for something added later we can deal with it then.

File: boards/modalai/voxl2/cmake/voxl2_posix.cmake (hunk 3 above)
Comment (katzfey) [reply boards/modalai/voxl2/cmake/voxl2_posix.cmake:53]: Nope, I'll remove them from this file

File: cmake/kconfig.cmake (hunk 2 above)
Comment (katzfey) [reply cmake/kconfig.cmake:232]: The voxl2_posix.cmake definitely wasn't needed there so I removed it. Still looking at best options for qurt though.

File: boards/modalai/voxl2/cmake/voxl2_posix.cmake (hunk 5 above)
Comment (katzfey) [reply boards/modalai/voxl2/cmake/voxl2_posix.cmake]: I was able to get rid of this and clean up the voxl2 board specifics in the posix platform cmake file

File: boards/modalai/voxl2/cmake/voxl2_posix.cmake [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+	"${PX4_SOURCE_DIR}/platforms/posix/cmake"
+)
+
+# set(DISABLE_PARAMS_MODULE_SCOPING TRUE)
+
+add_definitions(-DORB_COMMUNICATOR)
+
+set(CONFIG_PARAM_SERVER "1")
+
+add_definitions( -D__PX4_LINUX )
```
Comment (dagar) [inline boards/modalai/voxl2/cmake/voxl2_posix.cmake:52]:
```suggestion
```

This should already be added when you have `CONFIG_BOARD_LINUX=y` set.

File: boards/modalai/voxl2/cmake/voxl2_posix.cmake [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+list(APPEND CMAKE_MODULE_PATH
+	"${PX4_SOURCE_DIR}/platforms/posix/cmake"
+)
+
+# set(DISABLE_PARAMS_MODULE_SCOPING TRUE)
+
+add_definitions(-DORB_COMMUNICATOR)
+
+set(CONFIG_PARAM_SERVER "1")
```
Comment (dagar) [inline boards/modalai/voxl2/cmake/voxl2_posix.cmake:50]:
```suggestion
```

Doesn't exist yet?

File: boards/modalai/voxl2/cmake/voxl2_qurt.cmake [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+if ("$ENV{HEXAGON_TOOLS_ROOT}" STREQUAL "")
+	message(FATAL_ERROR "Environment variable HEXAGON_TOOLS_ROOT must be set")
+else()
+	set(HEXAGON_TOOLS_ROOT $ENV{HEXAGON_TOOLS_ROOT})
+endif()
+
+include(px4_git)
+
+list(APPEND CMAKE_MODULE_PATH
+	"${PX4_SOURCE_DIR}/platforms/qurt/cmake"
```
Comment (dagar) [inline boards/modalai/voxl2/cmake/voxl2_qurt.cmake:31]: This should already be happening automatically based on the platform.

https://github.com/PX4/PX4-Autopilot/blob/e21025e8f938bcd00eb86833495ba8a2efa65a41/cmake/kconfig.cmake#L227

File: platforms/qurt/cmake/qurt_reqs.cmake [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+	-Wl,--wrap=__stack_chk_fail
+	${DYNAMIC_LIBS}
+	-lc
+	${EXTRALDFLAGS}
+	)
+
+include (CMakeParseArguments)
+
+# Process DSP files
+function (QURT_LIB)
```
Comment (dagar) [inline platforms/qurt/cmake/qurt_reqs.cmake:182]: QURT_LIB seems like another one where we can just use regular cmake directly?

The deploy helper could be done generically for all libs (in the platform/qurt/ layer) or specifically for a few special targets?

File: boards/modalai/voxl2/cmake/voxl2_posix.cmake (hunk 8 above)
Comment (katzfey) [reply boards/modalai/voxl2/cmake/voxl2_posix.cmake:50]: Correct, only there for future expansion :-)

File: platforms/posix/CMakeLists.txt [hunk 11]
```diff
@@ -101,6 +101,14 @@ endif()
 if("${PX4_BOARD}" MATCHES "beaglebone_blue")
 	target_link_libraries(px4 PRIVATE robotics_cape)
 
+elseif("${PX4_BOARD}" MATCHES "modalai_voxl2")
```
Comment (dagar) [inline platforms/posix/CMakeLists.txt:104]: We should invent some mechanism that pushes this into the board directory.

(note for later)

File: cmake/kconfig.cmake [hunk 12]
```diff
@@ -228,6 +228,10 @@ if(EXISTS ${BOARD_DEFCONFIG})
 
         # platform-specific include path
         include_directories(${PX4_SOURCE_DIR}/platforms/${PX4_PLATFORM}/src/px4/common/include)
+
+        if(PLATFORM STREQUAL "qurt")
+            include(${PX4_SOURCE_DIR}/boards/modalai/voxl2/cmake/voxl2_qurt.cmake)
```
Comment (dagar) [inline cmake/kconfig.cmake:233]: Isn't this basically all a toolchain anyway?

voxl2_qurt.cmake + qurt_reqs.cmake + Toolchain-qurt.cmake => CMAKE_TOOLCHAIN_FILE?

See if(TOOLCHAIN) below on line 241.

**Review by dagar — APPROVED** [review]

File: cmake/px4_add_library.cmake (hunk 1 above)
Comment (katzfey) [reply cmake/px4_add_library.cmake:50]: Actually, uorb_headers does not need to be excluded. But parameters does. It opens a whole list of issues that will need to be dealt with when we get to the parameters module.
